modules/internal/DefaultRectangular.chpl

```diff
@@ -2174,6 +2174,69 @@ module DefaultRectangular {
   proc DefaultRectangularArr.isDefaultRectangular() param return true;
   proc type DefaultRectangularArr.isDefaultRectangular() param return true;
 
+  config param debugDRScan = false;
+
+  /* This computes a 1D scan in parallel on the array, for 1D arrays only */
+  proc DefaultRectangularArr.doiScan(op, dom) where (rank == 1) {
+    use RangeChunk;
+
+    type resType = op.generate().type;
+    var res: [dom] resType;
+
+    // Compute who owns what
+    // TODO: throttle tasks better than this
+    const numTasks = min(here.numPUs(), dom.dsiDim(1).size);
+    const rngs = RangeChunk.chunks(dom.dsiDim(1), numTasks);
+    if debugDRScan {
+      writeln("Using ", numTasks, " tasks");
+      writeln("Whose chunks are: ", rngs);
+    }
+
+    var state: [1..numTasks] resType;
```

Comment:
Should we add an "unimplemented" error when accumulation state and the result have different types, ex. if `resType != op.identity.type` ?

Otherwise users may get unclear resolution errors or possibly semantically-incorrect resolution outcomes.

Reply:
I'd wanted to do something like this (or to put the check into a `where` clause, but couldn't figure out how to get the state type.  I hadn't thought about identity, thanks for the tip!

Am I remembering correctly that we decided to make all of these types a bit more explicit in the new interface, or is that just wishful thinking on my part?  :)

Reply:
> we decided to make all of these types a bit more explicit in the new interface?

The new interface #9879 uses them explicitly in the types of the formals.

The reduce-op record is not required to provide the corresponding methods. Because they are not needed in the implementation afaik.